testing/trino-product-tests/src/main/java/io/trino/testng/services/VerifyProductTestGroups.java

```diff
@@ -0,0 +1,117 @@
+import static io.trino.tests.product.TestGroups.PROFILE_SPECIFIC_TESTS;
+import static java.lang.Boolean.getBoolean;
+import static java.lang.Boolean.parseBoolean;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * Verify that a product test has reasonably assigned groups. Incorrectly assigned groups
+ * may lead to a test executing too many times, or not executing at all.
+ */
+public class VerifyProductTestGroups
+        implements IInvokedMethodListener
+{
+    public static final boolean DISABLED = getBoolean("VerifyProductTestGroups.disabled") || parseBoolean(System.getenv("VERIFY_PRODUCT_TEST_GROUPS_DISABLED"));
+
+    // Lazy so that execution is covered with reportListenerFailure()
+    private final Supplier<Set<String>> definedTestGroups = memoize(() ->
+            Stream.of(TestGroups.class.getFields())
+                    .map(field -> {
+                        try {
+                            return (String) field.get(null);
+                        }
+                        catch (ReflectiveOperationException e) {
+                            throw new RuntimeException(e);
+                        }
+                    })
+                    .collect(toImmutableSet()));
+
+    @Override
+    public void beforeInvocation(IInvokedMethod method, ITestResult testResult)
+    {
+        try {
+            verify(method);
+        }
+        catch (Throwable e) {
+            reportListenerFailure(
+                    VerifyProductTestGroups.class,
+                    "Failed to process %s: \n%s",
+                    method,
+                    getStackTraceAsString(e));
+        }
+    }
+
+    private void verify(IInvokedMethod method)
+    {
+        if (DISABLED) {
+            return;
+        }
+        if (!method.isTestMethod()) {
+            return;
+        }
+
+        Method javaMethod = method.getTestMethod().getConstructorOrMethod().getMethod();
+        if (javaMethod != null && javaMethod.isAnnotationPresent(Suppress.class)) {
+            return;
+        }
+
+        String methodName = formatTestName(method.getTestResult());
+        Set<String> groups = ImmutableSet.copyOf(method.getTestMethod().getGroups());
+
+        groups.stream()
+                .filter(Predicate.not(definedTestGroups.get()::contains))
+                .findAny().ifPresent(group -> {
+                    throw new IllegalStateException("Test group %s on %s is not defined in TestGroups".formatted(group, methodName));
+                });
+
+        if (groups.contains(PROFILE_SPECIFIC_TESTS)) {
+            checkState(groups.size() > 1, "Test group %s on %s must be used with other groups", PROFILE_SPECIFIC_TESTS, methodName);
+        }
+        else {
+            checkState(groups.contains(DEFAULT_TEST_SUITE), "If %s is not in %s group, it should be marked with %s", methodName, PROFILE_SPECIFIC_TESTS, DEFAULT_TEST_SUITE);
```

Comment:
I didn't get this case. In cases where a test isn't a PROFILE_SPECIFIC_TEST I don't think one should need to add explicit `DEFAULT_TEST_SUITE`.

Should this check just be that at-least 1 non-PROFILE_SPECIFIC_TEST group is present?

Or do I misunderstand everything?